browser/brave_cosmetic_resources_tab_helper.cc

```diff
@@ -0,0 +1,455 @@
+        json_selectors.empty()) {
+      json_selectors = "[]";
+    }
+    std::string new_selectors_script =
+        base::StringPrintf(kStyleSelectorsInjectScript, json_selectors.c_str());
+    if (!json_selectors.empty()) {
+      auto* frame_host = content::RenderFrameHost::FromID(frame_id);
+      if (!frame_host)
+        return;
+      frame_host->ExecuteJavaScriptInIsolatedWorld(
+          base::UTF8ToUTF16(new_selectors_script), base::NullCallback(),
+          ISOLATED_WORLD_ID_CHROME_INTERNAL);
+    }
+  }
+
+  if (!enabled_1st_party_cf_filtering_) {
+    auto* frame_host = content::RenderFrameHost::FromID(frame_id);
+    if (!frame_host)
+      return;
+    frame_host->ExecuteJavaScriptInIsolatedWorld(
+        base::UTF8ToUTF16(*g_observing_script), base::NullCallback(),
+        ISOLATED_WORLD_ID_CHROME_INTERNAL);
+  }
+}
+
+void BraveCosmeticResourcesTabHelper::GetHiddenClassIdSelectorsOnUI(
+    content::GlobalFrameRoutingId frame_id,
+    const GURL& url,
+    std::unique_ptr<base::ListValue> selectors) {
+  if (!selectors || IsVettedSearchEngine(url.host())) {
+    return;
+  }
+  for (size_t i = 0; i < selectors->GetSize(); i++) {
+    base::ListValue* selectors_list = nullptr;
+    if (!selectors->GetList()[i].GetAsList(&selectors_list) ||
+        selectors_list->GetSize() == 0) {
+      continue;
+    }
+    std::string json_selectors;
+    if (!base::JSONWriter::Write(*selectors_list, &json_selectors) ||
+        json_selectors.empty()) {
+      json_selectors = "[]";
+    }
+    // Building a script for stylesheet modifications
+    std::string new_selectors_script =
+        base::StringPrintf(kSelectorsInjectScript, json_selectors.c_str());
+    if (selectors_list->GetSize() != 0) {
+      auto* frame_host = content::RenderFrameHost::FromID(frame_id);
+      if (!frame_host)
+        return;
+      frame_host->ExecuteJavaScriptInIsolatedWorld(
+          base::UTF8ToUTF16(new_selectors_script), base::NullCallback(),
+          ISOLATED_WORLD_ID_CHROME_INTERNAL);
+    }
+  }
+
+  if (!enabled_1st_party_cf_filtering_) {
+    auto* frame_host = content::RenderFrameHost::FromID(frame_id);
+    if (!frame_host)
+      return;
+    frame_host->ExecuteJavaScriptInIsolatedWorld(
+        base::UTF8ToUTF16(*g_observing_script), base::NullCallback(),
+        ISOLATED_WORLD_ID_CHROME_INTERNAL);
+  }
+}
+
+void BraveCosmeticResourcesTabHelper::ProcessURL(
+    content::RenderFrameHost* render_frame_host,
+    const GURL& url,
+    const bool do_non_scriptlets) {
+  content::CosmeticFiltersCommunicationImpl::CreateInstance(render_frame_host,
+                                                            this);
+  if (!render_frame_host || !ShouldDoCosmeticFiltering(web_contents(), url)) {
+    return;
+  }
+  g_brave_browser_process->ad_block_service()
+      ->GetTaskRunner()
+      ->PostTaskAndReplyWithResult(
+          FROM_HERE,
+          base::BindOnce(GetUrlCosmeticResourcesOnTaskRunner, url.spec()),
```

Comment:
since we post to another thread, nothing prevents our frame to navigate away in parallel, so we will run scripts on another page.
One day we really should move obtaining Cosmetic stuff to UI in ad block services.

Reply:
it's just too late, of course we cannot fetch things by url on UI, so will need to continue living with it